app/models/concerns/key_pair_compute_resource.rb

```diff
@@ -0,0 +1,55 @@
+module KeyPairComputeResource
+  extend ActiveSupport::Concern
+
+  included do
+    has_one :key_pair, :foreign_key => :compute_resource_id, :dependent => :destroy
+    after_create :setup_key_pair
+    after_destroy :destroy_key_pair
+    delegate :key_pairs, :to => :client
+  end
+
+  def get_compute_key_pairs
+    return unless capabilities.include?(:key_pair)
+    key_pairs = self.try(:key_pairs)
+    active_key = key_pair
+    return [] if key_pairs.nil? || active_key.nil?
+    compute_resource_key_pair_class = "ComputeResourceKeyPairs::#{provider_friendly_name}".constantize
+    key_pairs.map do |key|
+      compute_resource_key_pair_class.new(key.name, key.fingerprint, active_key.name, active_key.id)
+    end
+  end
+
+  def recreate
+    destroy_key_pair
+    setup_key_pair
+  end
+
+  def delete_key_pair(kay_pair_name)
```

Comment:
i wonder if all of the deleation of keys (that are not used by the compute resource) should be in the ComputeResourceKeyPairs::Base class.